### PR TITLE
fix(routing): partially dynamic segments are being truncated

### LINKED ISSUE #10122: Redirects for dynamic routes break when files with same name but different extensions exist
### Astro Info

```block
Astro                    v4.3.7
Node                     v20.11.0
System                   Linux (x64)
Package Manager          bun
Output                   static
Adapter                  none
Integrations             @astrojs/mdx
                         @astrojs/sitemap
```

### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

I am developing a blog based on the [Astro Starter Kit: Blog](https://github.com/withastro/astro/tree/main/examples/blog). The articles can be fetched at `/blog/[...slug]`, so I've made it possible to fetch the OGP images of the articles at `/blog/[...slug].png`. 

The problem starts here. I add a redirect from `/posts/[...slug]` to `/blog/[...slug]` in the Astro Config. However, when I access `/posts/[...slug]`, it redirects to `/blog/[...slug].png` instead of `/blog/[...slug]`.

### What's the expected result?

Access to `/posts/[...slug]` is redirected to `/blog/[...slug]`.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-ndv9vp

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by 1574242600**
## Changes
fix #10122
> partially dynamic segments are being truncated and affecting routing.
## Testing
units/routing/manifest.test.js

### CODE REVIEW

File: packages/astro/src/core/routing/manifest/create.ts [hunk 1]
```diff
@@ -899,3 +893,17 @@ function computeRoutePriority(config: AstroConfig): RoutePriorityOverride {
...
+
+function joinSegments(segments: RoutePart[][]): string {
+	const arr = segments.map(function d(segment): string {
+		const [{ dynamic, content }] = segment;
+
+		return segment.length > 1
+			? segment.map((s) => d([s])).join('')
+			: dynamic
+				? `[${content}]`
+				: content;
```
Comment (Fryuni) [inline packages/astro/src/core/routing/manifest/create.ts:905]: Spread segments should be `[...${content}]`

**Comment by Fryuni** [comment]
I reviewed it from my phone, so it wat just a quick look over. I'll review it more carefully later today.

Thanks for the PR!

File: .changeset/five-colts-rescue.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"astro": patch
+---
+
+Fix the issue of partially truncated dynamic segments.
```
Comment (florian-lefebvre) [inline .changeset/five-colts-rescue.md:5]:
```suggestion
Fixes a routing issue with partially truncated dynamic segments.
```
Would be great to add an example as well maybe?

File: packages/astro/src/core/routing/manifest/create.ts (hunk 1 above)
Comment (1574242600) [reply packages/astro/src/core/routing/manifest/create.ts:905]: `content` of rest parameter is `...xxx`,i'm not entirely certain whether it requires separate handling.

File: packages/astro/src/core/routing/manifest/create.ts [hunk 3]
```diff
@@ -899,3 +893,17 @@ function computeRoutePriority(config: AstroConfig): RoutePriorityOverride {
 	}
 	return 'legacy';
 }
+
+function joinSegments(segments: RoutePart[][]): string {
+	const arr = segments.map(function d(segment): string {
+		const [{ dynamic, content }] = segment;
+
+		return segment.length > 1
+			? segment.map((s) => d([s])).join('')
```
Comment (lilnasy) [inline packages/astro/src/core/routing/manifest/create.ts:902]: Could you confirm I have got this right?

This one line is the fix.

Previously, we ignored all but the first `RoutePart`.

Now, each `RoutePart` gets passed to this `d()` function. And the function ultimately still wraps the dynamic `RouteParts` in square brackets same as it was done before.

Comment (1574242600) [reply packages/astro/src/core/routing/manifest/create.ts:902]: Sure thing! You've got it right.
Would there be any issues with this?

Comment (lilnasy) [reply packages/astro/src/core/routing/manifest/create.ts:902]: No, this is great - just took me a while to understand the recursive map, so I thought I'd double check.

**Review by lilnasy — APPROVED** [review]
Looks great to me! Great work on the test as well.

File: packages/astro/test/units/routing/manifest.test.js [hunk 4]
```diff
@@ -144,8 +144,7 @@ describe('routing - createRouteManifest', () => {
 
 		assertRouteRelations(getManifestRoutes(manifest), [
 			['/', '/[...rest]'],
-			['/static', '/static-'],
-			['/static-', '/[dynamic]'],
```
Comment (Fryuni) [inline packages/astro/test/units/routing/manifest.test.js:148]: This check shouldn't be removed, it should no longer be truncated but still be there

**Review by Fryuni — CHANGES_REQUESTED** [review]
I think the recursive mapping is more confusing than needed and either two normal for loops or two nested maps would make it easier to comprehend.

There is one test check that was removed instead of updated.

Besides that, looking good! Sorry for the delay in the review

File: packages/astro/test/units/routing/manifest.test.js (hunk 4 above)
Comment (lilnasy) [reply packages/astro/test/units/routing/manifest.test.js:148]: It is combined into a single rule, isn't it?

Comment (Fryuni) [reply packages/astro/test/units/routing/manifest.test.js:148]: No, this is checking two different rules
- `/static` comes before `/static-[dynamic]` (fully static vs mixed static + dynamic)
- `/static-[dynamic]` comes before `/[dynamic]` (fully dynamic vs mixed static + dynamic)

**Review by bluwy — APPROVED** [review]

**Review by Fryuni — APPROVED** [review]
LGTM

**Review by ematipico — APPROVED** [review]

**Comment by arjendevos** [comment]
Issue still exists in version `"astro": "^4.7.1"`

Using it like this:
```ts
redirects: {
    "/nl/blog/[slug]": "/blog/[slug]",
  },
```

but it redirects to the `.png` version

**Pull request merged**